---
title: "Validating hpexome tool"
date: "`r Sys.Date()`"
output: github_document
---

Command line to execute hpexome on NA12877_S1 sample

```bash
REF_DIR=~/bioinf/ref/hg19
hpexome \
    --bam NA12877_S1.bam \
    --genome ucsc.hg19.fasta \
    --dbsnp dbsnp_138.hg19.vcf \
    --indels ALL.wgs.indels_mills_devine_hg19_leftAligned_collapsed_double_hit.indels.sites.vcf.gz \
    --indels ALL.wgs.low_coverage_vqsr.20101123.indels.sites.vcf.gz \
    --sites ALL.wgs.dbsnp.build135.snps.sites.vcf.gz \
    --stand_call_conf 40 \
    --stand_emit_conf 10 \
    --min_pruning 3 \
    --scatter_count 16 \
    --job_runner PbsEngine \
    hpexome_validation
```

Packages
```{r message=FALSE}
library(VariantAnnotation)
library(knitr)
library(ggplot2)
```


Normalize VCF files:
```{r}
workDir <- "/home/bioinf/hpexome_validation"
genomeFile <- file.path(workDir, "genome.chr.fa")
NA12877.ref <- file.path(workDir, "NA12877.vcf.gz")
NA12877.ref.vt <- file.path(workDir, "NA12877.vt.vcf")
NA12877.vcf <- file.path(workDir, "validate_NA12877_try1/NA12877_S1.sort.fix.HC.raw.vt.vcf.bz")
NA12877.vcf.vt <- file.path(workDir, "validate_NA12877_try1/NA12877_S1.sort.fix.HC.raw.vt.vcf")
```

    `r paste("vt decompose", NA12877.ref, "| vt normalize -r", genomeFile, "- | vt uniq - -o", NA12877.ref.vt)`
    `r paste("vt decompose", NA12877.vcf, "| vt normalize -r", genomeFile, "- | vt uniq - -o", NA12877.vcf.vt)`

```{r}
NA12877.ref.vt.bz <- paste0(NA12877.ref.vt, ".bz")
# bgzip(NA12877.ref.vt, NA12877.ref.vt.bz)
NA12877.vcf.vt.bz <- paste0(NA12877.vcf.vt, ".bz")
# bgzip(NA12877.ref.vt, NA12877.vcf.vt.bz)
```

Load VCF files into R:
```{r}
genome <- "hg19"
ref <- readVcf(NA12877.ref.vt.bz, genome)
vcf <- readVcf(NA12877.vcf.vt.bz, genome)
```

Filter VCFs by selected chromosomes
```{r}
chr.select <- paste0("chr", 1:22)
seqlevels(ref, force=TRUE) <- seqlevels(ref)[seqlevels(ref) %in% chr.select]
seqlevels(vcf, force=TRUE) <- seqlevels(vcf)[seqlevels(vcf) %in% chr.select]
```

Remove variants from reference that have genotype "0/."
```{r}
idx <- which(!grepl("0(\\/|\\|)\\.", geno(ref)$GT))
ref <- ref[idx, ]
```

Number of variants:
```{r}
kable(data.frame(reference=nrow(ref), hpexome=nrow(vcf)))
```

Subset VCF files using common variants between the two VCFs:
```{r}
vcf.ranges <- rowRanges(vcf)
ref.ranges <- rowRanges(ref)
overlap <- findOverlaps(vcf.ranges, ref.ranges)
vcf.sub <- vcf[queryHits(overlap), ]
ref.sub <- ref[subjectHits(overlap), ]
```

Number of variants after subset (must be the same value for the two file):
```{r}
kable(data.frame(reference=nrow(ref.sub), hpexome=nrow(vcf.sub)))
```

The hpexome tool not found `r nrow(ref) - nrow(ref.sub)` variants.
However this tool found `r nrow(vcf) - nrow(vcf.sub)` new variants.
Because hpxome found `r nrow(vcf) - nrow(ref)` more variants than reference.

Create genotype count table
```{r}
vcf.gt <- geno(vcf.sub)$GT
ref.gt <- sub("\\|", "/", geno(ref.sub)$GT)
ref.gt <- sub("1\\/0", "0/1", ref.gt)

genoCount <- table(ref.gt, vcf.gt)
genoCount
```

Degree of concordance
```{r}
sum(diag(genoCount)) / sum(genoCount) * 100
```

Create intevals for `QG` value:
```{r}
gq <- geno(vcf.sub)$GQ
breaks <- unique(quantile(gq, seq(0, 1, .01), na.rm=TRUE))
intervals <- cut(gq, breaks)
```

This function count the calculates the number of correct genotypes:
```{r}
f <- function(interval, vcf.gt, ref.gt, intervals)
{
  idx <- interval == intervals
  sum(vcf.gt[idx] == ref.gt[idx], na.rm=TRUE) / length(ref.gt[idx]) * 100
}
```

Create table for intervals and correct genotypes:
```{r}
correct <- sapply(levels(intervals), f, vcf.gt, ref.gt, intervals, USE.NAMES = FALSE)
df <- data.frame(intervals=levels(intervals), correct)
kable(df)
```

Plot intervals
```{r}
ggplot(df, aes(x=intervals, y=correct)) +
    geom_bar(stat="identity") +
    labs(x="QG value intervals", y="% correct variant")
```
